### ISSUE TITLE
Bug with calculate_state

### ISSUE THREAD

**Issue by adam-a-a** [watertap-org/watertap#1123]
### Description

using calculate_state with volumetric flow and mass concentration sometimes results in erroneous results for flow_mol_phase_comp

### Expected Behavior

using calculate_state with volumetric flow and mass concentration should not result in erroneous results for flow_mol_phase_comp

### Steps to Reproduce

@hunterbarber @kurbansitterley please point me to your sections of code where you implement this, as well as "problematic" flow_vol and conc_mass values where erroneous results for flow_mol were observed.

**Comment by adam-a-a** [comment]
I have updates on this.
**TLDR:** `calculate_state` can yield erroneous values for state variables (e.g., `flow_mass_phase_comp` or `flow_mol_phase_comp`) **when providing very small values for, e.g., mass or molar concentrations.** The problem with this is that the user can easily miss that the solution is incorrect because `assert_optimal_termination` would pass. This issue is not just for MCAS but likely for all usages of `calculate_state` in property models.

**More details:**
We knew that we should calculate scaling factors before calling `calculate_state`, but what wasn't apparent was that if the user didn't provide scaling factors for very small values, or the scaling factors weren't sufficient enough, `calculate_state `could converge to a supposedly "optimal" solution because the small variable values provided by the user would lead to constraints being satisfied within tolerance. Technically, we knew this to be a potential caveat in general (i.e., very small values can "trick" the solver into "thinking" it converged to a correct solution)--we just didn't realize that this was slipping by calculate_state in particular.

**Remedy?**
Of course, one would think that scaling your property variables appropriately before running calculate_state would resolve the issue, and that's likely the case in most instances. The problematic scenario that I see here concerns the user who thinks they scaled appropriately but didn't, and now they are getting back their calculated state variables, which are incorrect, without realizing it. I am curious to hear thoughts on this, but one safeguard could be to add an optional argument to `calculate_state` that when true, would automatically scale the property variables specified in `var_args` (could be dangerous). Either that, or we can add default scaling factors for `var_args `in the case that they are missing scaling factors (e.g., scaling_factor= 10 /value) and log a warning. For example (omitting details, e.g., variable indices):

```
m.fs.properties_in.calculate_state(var_args = {'flow_vol': 1, 'conc_mass': 1e-12},
                                                                         hold_state=True}

#within calculate_state, scan the specified vars and check if they were given scaling factors. If not, set defaults.
for var in var_args.keys():
    model_var = getattr(self, var)
    if iscale.get_scaling_factor(model_var) is None:
        iscale.get_scaling_factor(model_var, default=value(10 / model_var), warning=True)
...
#towards the end of the method, assuming we flagged vars that were provided default scaling factors, we could unset their scaling factors afterwards
```
@andrewlee94 @bknueven @kurbansitterley @hunterbarber @TimBartholomew @avdudchenko  - if you (or anyone else) have any thoughts, let me know what you think before I attempt to address (or ignore) this.

**Comment by bknueven** [comment]
The safest thing would be, when possible, to use `calculate_variable_from_constraint`, instead of calling Ipopt. Not sure if that applies in this case. A better initialization could resolve some of these issues also.

In general, we can't really solve to tighter tolerances unless we "lift the hood" on Ipopt. There's a tread working on this, but obviously it's a longer process.

**Comment by adam-a-a** [comment]
OK, so instead of using `solve_indexed_blocks` which is what is used now within `calculate_state`, you're suggesting to swap that out and use `calculate_variable_from_constraint`, correct?

**Comment by bknueven** [comment]
Well, if it's possible. It likely isn't, unless the calculation is completely forward.

In this particular case, you could try passing in the `optarg` and specifying a tighter tolerance for ipopt, e.g., `optarg={"tol":1e-12}`. Note the default in WaterTAP is `tol=1e-8`.

Either way, as we don't use exact arithmetic, the user either needs to know appropriate scaling factors or an appropriate tolerance. Even if we tighten the defaults (which we might be able to do someday) there will still be a limit.

**Comment by adam-a-a** [comment]
Forgot to mention that this is partially tied to PR #1136

**Comment by TimBartholomew** [comment]
I think we should return an error if we find that the user hasn't touched the properties and hasn't ran calculate_scaling_factors before calling calculate_state
